defi-wonderland/balancer-v1-amm
# feat: helper support for unbalanced pool

These changes make the helper support two-token pools with arbitrary weights.

More discussion on the math can be found [here](https://cowservices.slack.com/archives/C035TDPHQAJ/p1719567592184189?thread_ts=1719410711.653979&cid=C035TDPHQAJ) (I hope this link works for your team!). Message content copied below for accessibility.

More details are available as inline comments.

Notably, the `GetTradeableOrder` library is not needed anymore, since the math is now specific to this AMM version.

### How to test

Unit/integration tests have been changed to account for the new pool format. Logic changes are needed because the definition of the spot price isn't just the rate of the two pool balances when the weights are different.

Note that the "unreasonable amount" check has been removed. This is because it's hard to succinctly come up with a reasonableness check that works for both equal-weight and distinct-weight pool and is menaningfully readable. I don't think this is an issue since in both tests involving this check we also verify that the final price of the AMM coincides with `EXPECTED_FINAL_SPOT_PRICE`, which is reasonable by construction.



<details><summary>Message with rationale for the math.</summary>

you want to calculate the sell amount so that, if the AMM sells tokens at the oracle price then its spot price equals the oracle price. So, if the oracle price is p, and you start with `b_x/w_x>p b_y/w_y`, you want to sell x for x/p such that

```
(b_x-x)/w_x=p (b_y+x/p)/w_y .
```

To put it another way, the calculations are correct if, in the end, the AMM spot price equals the Oracle. Also, the calculations have nothing to do with the "buy amount."

To conclude the calculations: you can solve the equation and get:

```
x = (w_x * w_y/(w_x+w_y))*(b_x/w_x - P (b_y/w_y))
```

note that the formula as in the reference implementation is

```
x = (b_x/2 - P (b_y/2))
```

so if you want to recycle it you need to normalize `w_x+w_y=2`  and redefine the pool reserves as  `b_x*w_y`  and `b_y*w_x`.
Then, once the calculations for the sell amount are correct, you just put its value in the `inGivenOut` formula to get the "buyAmount"

</details>
